src/collection-view.js

```diff
@@ -101,12 +101,21 @@ Marionette.CollectionView = Marionette.View.extend({
   _onCollectionAdd: function(child, collection, opts) {
     // `index` is present when adding with `at` since BB 1.2; indexOf fallback for < 1.2
     var index = opts.at !== undefined && (opts.index || collection.indexOf(child));
+    var models = collection.models;
 
     // When filtered or when there is no initial index, calculate index.
-    if (this.getOption('filter') || index === false) {
-      index = _.indexOf(this._filteredSortedModels(index), child);
+    if (this.getOption('filter') || this.getViewComparator() || index === false) {
+      models = this._filteredSortedModels(index);
     }
 
+    // find adjusted index
+    _.reduce(models, function(counter, model) {
+      var view = this.children.findByModel(model);
+      if (view) { return view._index; }
+
+      return model === child ? index = counter + 1 : counter;
+    }, -1, this);
```

Comment:
Hey @scott-w,

The reduce is used just to hold the counter as the models are iterated through. As for what is happening internally:

On every iteration we attempt to find the view corresponding to the iterated model. The `_index` of that view is then set as our current counter value so that when we finally hit our newly added model, we have an accurate representation of where it should lie relative to other rendered views.